src/main/java/org/eclipsefoundation/marketplace/service/impl/DefaultPromotionService.java

```diff
@@ -0,0 +1,171 @@
+		// for caching
+		Map<String, List<String>> adds = new HashMap<>();
+		adds.put("type", Arrays.asList("Listing"));
+		adds.put(UrlParameterNames.IDS.getParameterName(),
+				promos.stream().map(Promotion::getListingId).collect(Collectors.toList()));
+
+		MongoQuery<Listing> q = new MongoQuery<>(null, adds, listingFilter);
+		// retrieve the possible cached object
+		Optional<List<Listing>> cachedResults = listingCache.get("promo|listings", wrapper, adds,
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent()) {
+			LOGGER.error("Error while retrieving cached promotion listings");
+			return Collections.emptyList();
+		}
+
+		// return the results as a response
+		return cachedResults.get();
+	}
+
+	@Override
+	public void injectPromotions(RequestWrapper wrapper, List<Listing> listings) {
+		// create an empty promo query to get all promos
+		MongoQuery<Promotion> q = new MongoQuery<>(null, Collections.emptyMap(), promotionFilter);
+		// retrieve the possible cached object
+		Optional<List<Promotion>> cachedResults = promoCache.get("all|promo", wrapper, Collections.emptyMap(),
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent() || cachedResults.get().isEmpty()) {
+			LOGGER.debug("Could not find any promotions to inject, returning");
+			return;
+		}
+		// make a copy of the array to not impact cached values
+		List<Promotion> promos = new ArrayList<>(cachedResults.get());
+		List<Promotion> promoHolding = new ArrayList<>(promoCount);
+		LOGGER.debug("Found {} promotions, maximum number to inject {}", promos.size(), promoCount);
+		// check each promotion to see if it should be injected
+		Promotion curr;
+		while ((curr = getWeightedPromotion(promos)) != null && promoHolding.size() <= promoCount) {
+			// create a local final field referencing the current promotion for stream ref
+			final Promotion p = curr;
+			LOGGER.debug("Checking promo {}", p.getListingId());
+			// check if current promo matches any of the listing IDs
+			if (promos.stream().noneMatch(l -> l.getId().equals(p.getListingId()))) {
+				LOGGER.debug("Preparing promo with listing ID '{}' to be injected into result set",
+						curr.getListingId());
+
+				promoHolding.add(curr);
+			}
+		}
+
+		// if we couldn't find enough promos, reinsert records
+		if (promoHolding.isEmpty()) {
+			LOGGER.debug("Could not find any promos to inject");
+		} else {
+			for (Listing listing : getListingsForPromotions(wrapper, promoHolding)) {
+				LOGGER.debug("Injecting promo with listing ID '{}' ", listing.getId());
+				listing.setPromotion(true);
+				listings.add(0, listing);
+			}
+		}
+	}
+
+	/**
+	 * Using the weighting set in the promotions (or default if not set), retrieve a
+	 * random result, taking weighting into account.
+	 * 
+	 * @param promos list of promotions to retrieve a result from. This list will be
+	 *               modified as part of this call to shuffle and pop the chosen
```

Comment:
This is a bad contract. Never expect clients to read the doc :). Program defensively and not modify your inputs. Do a copy as the first step in your method and remove this contract from the doc.

Reply:
I agree that unless you make it explicit (like putting terms in your signature) this can be a dangerous bit of code, even if its private. Would naming it `popWeightedPromo()` be a suitable compromise?